tools/java/org/unicode/cldr/api/XmlDataSource.java

```diff
@@ -0,0 +1,329 @@
+package org.unicode.cldr.api;
+
+import com.google.common.base.CharMatcher;
+import com.google.common.collect.HashMultiset;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedMap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multiset;
+import org.unicode.cldr.util.CachingEntityResolver;
+import org.xml.sax.Attributes;
+import org.xml.sax.ErrorHandler;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXParseException;
+import org.xml.sax.XMLReader;
+import org.xml.sax.helpers.DefaultHandler;
+import org.xml.sax.helpers.XMLReaderFactory;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/** Serializes a set of LDML XML files as a sequence of {@code CldrValue}s. */
+final class XmlDataSource extends AbstractDataSource {
+    private final CldrDataType dtdType;
+    private final ImmutableSet<Path> xmlFiles;
+    private final CldrDraftStatus minimalDraftStatus;
+    private final Function<Path, Reader> openFn;
+
+    // Memoized data map to avoid loading/parsing files more than once.
+    private volatile ImmutableMap<CldrPath, CldrValue> pathValueMap = null;
+    // Whether the memoized map is in DTD order (we re-sort after it's cached).
+    private volatile boolean isDtdOrder = false;
+    // Lock to protect state of memoized map and ordering flag. These fields are volatile because
+    // we're using "double checked locking" to read the cached map.
+    // See: https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java
+    private final Object cacheLock = new Object();
+
+    // TODO: Handle draft status properly (it's a METADATA attribute).
```

Comment:
In the canonical format, the draft status is always only on leaf nodes. We allowed for them at higher levels just to enable us to bulk import a whole subtree with a particular draft status. But once canonicalized, that doesn't occur. And shouldn't occur in any release.

Reply:
Ahh, turns out this isn't true:
common/transforms/Azerbaijani-Latin-BGN.xml
where the <transform> element contains the draft status rather than the <tRule> child element. I'll leave the code as-is for now since it matches the current behaviour and we can revisit once the expected semantics have been sorted out.

Reply:
^ is it something like the leaf 'distinguishing' element?

Reply:
Maybe? I have no idea though, so I'll wait until someone can clarify. So far I've never had to care about "distinguishing" vs "non-distinguishing" elements, so anything done in that regard would be new territory. Can you point at any documentation on the semantic meaning of "distinguishing" vs "non-distinguishing" elements??